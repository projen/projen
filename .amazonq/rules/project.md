# Projen-managed Project Instructions

This project is managed by [projen](https://github.com/projen/projen), a project configuration management tool.

## Important Guidelines

### Task Execution

- **Always use projen for task execution**: Run tasks using `node ./projen.js <task-name>` instead of directly using npm, yarn, or other package managers.
- **Check available tasks**: Look in `.projen/tasks.json` to see all available tasks, their descriptions, and steps.
- **Common tasks**:
  - `node ./projen.js` - Synthesize project configuration files
  - `node ./projen.js build` - Builds the project, including running tests
  - `node ./projen.js test` - Runs tests only
  - `node ./projen.js compile` - Compiles the source code only

### File Modifications

- **DO NOT manually edit generated files**: Files marked with a comment like "~~ Generated by projen. To modify..." should never be edited directly.
- **Modify configuration in .projenrc**: To change project configuration, always edit the `.projenrc.ts`, `.projenrc.py` or `.projenrc.json` etc. file and then run `node ./projen.js` to regenerate the project files.
- **Check .projenrc first**: Before suggesting changes to package.json, tsconfig.json, or other configuration files, always check if these are managed by projen and suggest changes to .projenrc instead.

### Dependencies

- **Add dependencies through projen**: Use the projen configuration to add dependencies instead of manually editing package.json or using npm/yarn install directly.
- **Example**: In .projenrc, use methods like `addDeps()`, `addDevDeps()`, or `addPeerDeps()` to add dependencies.

### Workflow

1. Make changes to .projenrc configuration file
2. Run `node ./projen.js` to synthesize and update generated files
3. Review the changes
4. Commit both .projenrc and the generated files

## Projen Configuration

This project's configuration is defined in the .projenrc file at the root of the repository. All project metadata, dependencies, scripts, and tooling configuration should be managed through this file.

## Additional Resources

- [Projen Documentation](https://projen.io)
- [Projen GitHub Repository](https://github.com/projen/projen)

# Development Best Practices

- **Always run build after changes**: After modifying any source or test file, run `node ./projen.js build` to ensure your changes compile and pass all tests.
- **Task completion criteria**: A task is not considered complete until:
  - All tests pass (`node ./projen.js test`)
  - There are no compilation errors (`node ./projen.js compile`)
  - There are no linting errors (usually part of the build, if not, run the linter defined in tasks.json)
  - The full build succeeds (`node ./projen.js build`)
