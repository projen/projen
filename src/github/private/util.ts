export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
