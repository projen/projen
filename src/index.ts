export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
