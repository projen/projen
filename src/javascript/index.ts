export * from "./bundler";
export * from "./eslint";
export * from "./jest";
export * from "./license-checker";
export * from "./node-package";
export * from "./node-project";
export * from "./npm-config";
export * from "./prettier";
export * from "./projenrc";
export * from "./typescript-config";
export * from "./upgrade-dependencies";
export * from "./yarnrc";
