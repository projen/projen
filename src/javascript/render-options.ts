import * as inventory from "../inventory";
import { InitProjectOptionHints } from "../option-hints";
import { ModuleImports } from "./private/modules";

const PROJEN_NEW = "__new__";
const TAB = makePadding(2);

/**
 * Options for `renderProjectOptions`.
 */
export interface RenderProjectOptions {
  /**
   * The project type to render.
   */
  readonly type: inventory.ProjectType;

  /**
   * Project arguments as passed to `projen new`.
   */
  readonly args: Record<string, any>;

  /**
   * Include commented out options.
   * @default InitProjectOptionHints.FEATURED
   */
  readonly comments?: InitProjectOptionHints;

  /**
   * Inject a `__new__` attribute to the project constructor with a stringified
   * version of the project parameters and a `jsiiFqn` attribute that includes
   * the FQN of the project type. This is needed in order to generate initial
   * projenrc files.
   *
   * @default false
   */
  readonly bootstrap?: boolean;

  /**
   * A list of fields to omit from the initial projenrc file.
   * @default - none
   */
  readonly omitFromBootstrap?: string[];

  /**
   * Prefix all imports with this string and the full module name
   * This is required when executing options code in a vm
   *
   * @default - only use submodule as prefix
   */
  readonly prefixImports?: string;
}

/**
 * Information passed from `projen new` to the project object when the project
 * is first created. It is used to generate projenrc files in various languages.
 */
interface ProjenInit {
  /**
   * The JSII FQN of the project type.
   */
  readonly fqn: string;

  /**
   * Initial arguments passed to `projen new`.
   */
  readonly args: Record<string, any>;

  /**
   * Include commented out options. Does not apply to projenrc.json files.
   */
  readonly comments: InitProjectOptionHints;
}

/**
 * Renders options as if the project was created via `projen new` (embeds the __new__ field).
 */
export function renderProjenInitOptions(
  fqn: string,
  args: Record<string, any>,
  comments: InitProjectOptionHints = InitProjectOptionHints.NONE
): any {
  return {
    ...args,
    [PROJEN_NEW]: { fqn, args, comments } as ProjenInit,
  };
}

export function resolveInitProject(opts: any) {
  const f = opts[PROJEN_NEW] as ProjenInit;
  if (!f) {
    return undefined;
  }

  const type = inventory.resolveProjectType(f.fqn);
  if (!type) {
    throw new Error(`unable to resolve project type for ${f.fqn}`);
  }
  return {
    args: f.args,
    fqn: f.fqn,
    type: type,
    comments: f.comments,
  };
}

/**
 * Prints all parameters that can be used in a project type, alongside their descriptions.
 *
 * Parameters in `params` that aren't undefined are rendered as defaults,
 * while all other parameters are rendered as commented out.
 *
 * Returns the printed output and a set of required imports as an object
 * in the form { options, imports }.
 */
export function renderJavaScriptOptions(opts: RenderProjectOptions): {
  renderedOptions: string;
  imports: ModuleImports;
} {
  const renders: Record<string, string> = {};
  const optionsWithDefaults: string[] = [];
  const allImports = new ModuleImports();

  for (const option of opts.type.options) {
    if (option.deprecated) {
      continue;
    }

    const optionName = option.name;

    if (opts.args[optionName] !== undefined) {
      const arg = opts.args[optionName];
      const { js, moduleName, importName } = renderArgAsJavaScript(arg, option);

      renders[optionName] = `${optionName}: ${js},`;
      optionsWithDefaults.push(optionName);

      if (moduleName && importName) {
        allImports.from(moduleName, importName);
        if (opts.prefixImports) {
          const prefix = `${opts.prefixImports}["${moduleName}"].`;
          renders[optionName] = `${optionName}: ${prefix}${js},`;
        }
      }
    } else {
      const defaultValue = option.default?.startsWith("-")
        ? undefined
        : option.default ?? undefined;
      renders[optionName] = `// ${optionName}: ${defaultValue},`;
    }
  }

  const bootstrap = opts.bootstrap ?? false;
  if (bootstrap) {
    for (const arg of opts.omitFromBootstrap ?? []) {
      delete opts.args[arg];
    }
    renders[PROJEN_NEW] = `${PROJEN_NEW}: ${JSON.stringify({
      args: opts.args,
      fqn: opts.type.fqn,
      comments: opts.comments,
    } as ProjenInit)},`;
    optionsWithDefaults.push(PROJEN_NEW);
  }

  // generate rendering
  const result: string[] = [];
  result.push("{");

  // render options with defaults
  optionsWithDefaults.sort();
  for (const optionName of optionsWithDefaults) {
    result.push(`${TAB}${renders[optionName]}`);
  }
  if (result.length > 1) {
    result.push("");
  }

  // render options without defaults as comments
  if (opts.comments === InitProjectOptionHints.ALL) {
    const options = opts.type.options.filter(
      (opt) => !opt.deprecated && opts.args[opt.name] === undefined
    );
    result.push(...renderCommentedOptionsByModule(renders, options));
  } else if (opts.comments === InitProjectOptionHints.FEATURED) {
    const options = opts.type.options.filter(
      (opt) =>
        !opt.deprecated && opts.args[opt.name] === undefined && opt.featured
    );
    result.push(...renderCommentedOptionsInOrder(renders, options));
  } else if (opts.comments === InitProjectOptionHints.NONE) {
    // don't render any extra options
  }

  if (result[result.length - 1] === "") {
    result.pop();
  }
  result.push("}");
  return { renderedOptions: result.join("\n"), imports: allImports };
}

function renderCommentedOptionsByModule(
  renders: Record<string, string>,
  options: inventory.ProjectOption[]
) {
  const optionsByModule: Record<string, inventory.ProjectOption[]> = {};

  for (const option of options) {
    const parentModule = option.parent;
    optionsByModule[parentModule] = optionsByModule[parentModule] ?? [];
    optionsByModule[parentModule].push(option);
  }

  for (const parentModule in optionsByModule) {
    optionsByModule[parentModule].sort((o1, o2) =>
      o1.name.localeCompare(o2.name)
    );
  }

  const result = [];
  const marginSize = Math.max(
    ...options.map((opt) => renders[opt.name].length)
  );
  for (const [moduleName, optionGroup] of Object.entries(
    optionsByModule
  ).sort()) {
    result.push(`${TAB}/* ${moduleName} */`);
    for (const option of optionGroup) {
      const paramRender = renders[option.name];
      const docstring = option.docs || "No documentation found.";
      result.push(
        `${TAB}${paramRender}${makePadding(
          marginSize - paramRender.length + 2
        )}/* ${docstring} */`
      );
    }
    result.push("");
  }
  return result;
}

function renderCommentedOptionsInOrder(
  renders: Record<string, string>,
  options: inventory.ProjectOption[]
) {
  const result = [];
  const marginSize = Math.max(
    ...options.map((opt) => renders[opt.name].length)
  );
  for (const option of options) {
    const paramRender = renders[option.name];
    const docstring = option.docs || "No documentation found.";
    result.push(
      `${TAB}${paramRender}${makePadding(
        marginSize - paramRender.length + 2
      )}/* ${docstring} */`
    );
  }
  return result;
}

/**
 * Renders a value as a JavaScript value, converting strings to enums where
 * appropriate. The type must be JSON-like (string, number, boolean, array,
 * enum, or JSON object).
 *
 * Returns a JavaScript expression as a string, and the names of any
 * necessary imports.
 */
function renderArgAsJavaScript(arg: any, option: inventory.ProjectOption) {
  if (option.kind === "enum") {
    if (!option.fqn) {
      throw new Error(`fqn field is missing from enum option ${option.name}`);
    }
    const parts = option.fqn.split("."); // -> ['projen', 'web', 'MyEnum']
    const enumChoice = String(arg).toUpperCase().replace(/-/g, "_"); // custom-value -> CUSTOM_VALUE
    const js = `${parts.slice(1).join(".")}.${enumChoice}`; // -> web.MyEnum.CUSTOM_VALUE
    const moduleName = parts[0]; // -> projen
    const importName = parts[1]; // -> web
    return { js, moduleName, importName };
  } else if (option.jsonLike) {
    return { js: JSON.stringify(arg) };
  } else {
    throw new Error(
      `Unexpected option ${option.name} - cannot render a value for this option because it does not have a JSON-like type.`
    );
  }
}

function makePadding(paddingLength: number): string {
  return " ".repeat(paddingLength);
}
