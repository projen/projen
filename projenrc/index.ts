import { DependencyType, JsonFile, Project, TextFile } from "../src";
import { PROJEN_MARKER } from "../src/common";
import { TaskWorkflow } from "../src/github";
import {
  NodeProject,
  UpgradeDependencies,
  UpgradeDependenciesSchedule,
} from "../src/javascript";

export * from "./windows-build";

/**
 * Add integration tests tasks to a project
 *
 * @param project The project to add the tasks to
 */
export function setupIntegTest(project: NodeProject) {
  const packagePythonTask = project.tasks.tryFind("package:python");

  const pythonCompatTask = project.addTask("integ:python-compat", {
    exec: "scripts/python-compat.sh",
    description:
      "Checks that projen's submodule structure does not cause import failures for python. Expects python to be installed and projen to be fully built.",
  });
  const integTask = project.addTask("integ", {
    description: "Run integration tests",
  });
  integTask.spawn(project.compileTask);
  if (packagePythonTask) {
    integTask.spawn(packagePythonTask);
  }
  integTask.spawn(pythonCompatTask);

  project.buildWorkflow?.addPostBuildJobTask(integTask, {
    tools: { python: { version: "3.x" }, go: { version: "1.16.x" } },
  });
}

/**
 * Add a task to bundle the run-task script needed for "projen eject"
 *
 * @param project The project to add the task to
 */
export function setupBundleTaskRunner(project: Project) {
  // build `run-task` script needed for "projen eject" functionality
  // TODO: use project.bundler.addBundle instead - currently it's too inflexible on where the output goes
  const taskRunnerPath = "lib/run-task.cjs";
  const task = project.addTask("bundle:task-runner", {
    description: 'Bundle the run-task.cjs script needed for "projen eject"',
    exec:
      `esbuild src/task-runtime.ts ` +
      `--outfile=${taskRunnerPath} ` +
      `--bundle ` +
      `--platform=node ` +
      `--format=cjs ` +
      `--external:"*/package.json" ` +
      `--banner:js="#!/usr/bin/env node" ` +
      `--footer:js="const runtime = new TaskRuntime(\\".\\");\nruntime.runTask(process.argv[2]);"`,
  });
  project.postCompileTask.spawn(task);
}

/**
 * Setup the projen bootstrap script
 *
 * @param project The project to add the tasks to
 * @param bootstrapScriptFile The path to the bootstrap script
 */
export function setupProjenBootstrap(
  project: NodeProject,
  bootstrapScriptFile: string
) {
  // this script is what we use as the projen command in this project
  // it will compile the project if needed and then run the cli.
  const bootstrapScript = new TextFile(project, bootstrapScriptFile, {
    executable: true,
    marker: true,
    lines: `#!/usr/bin/env node
// ${PROJEN_MARKER}

const { existsSync } = require("fs");
const { execSync } = require("child_process");

function execCommand(command) {
  try {
    execSync(command, { stdio: "inherit" });
  } catch (error) {
    console.error(\`Failed to execute command: \${command}\`, error);
    process.exit(1);
  }
}

const isBuild = existsSync("lib/cli/index.js");
const hasJsii = existsSync("node_modules/.bin/jsii");
const hasTsNode = existsSync("node_modules/.bin/ts-node");
const needsBootstrapping = !isBuild || !hasTsNode;

const installCommand = "yarn install --frozen-lockfile --check-files --non-interactive";
const buildCommand = "npx jsii --silence-warnings=reserved-word --no-fix-peer-dependencies";

function bootstrap() {
  console.info("bootstrapping...");

  if (!hasTsNode || !hasJsii) {
    execCommand(installCommand);
  }

  if (!isBuild) {
    execCommand(buildCommand);
  }
}

if (needsBootstrapping) {
  bootstrap();
}

const args = process.argv.slice(2).join(" ");
execCommand(\`node bin/projen \${args}\`);
`.split("\n"),
  });

  project.npmignore?.exclude(`/${bootstrapScript.path}`);
}

/**
 * Add a task to check licenses
 *
 * @param project The project to add the task to
 */
export function setupCheckLicenses(project: NodeProject) {
  if (project.github) {
    const installCiTask = project.tasks.tryFind("install:ci");
    const checkLicensesTask = project.tasks.tryFind("check-licenses");

    if (installCiTask && checkLicensesTask) {
      // Run license checker as a separate CI job
      new TaskWorkflow(project.github, {
        name: "check-licenses",
        jobId: "check-licenses",
        triggers: {
          pullRequest: {},
          workflowDispatch: {},
          mergeGroup: {},
          push: { branches: ["main"] },
        },
        permissions: {},
        preBuildSteps: [
          {
            name: "Install dependencies",
            run: project.runTaskCommand(installCiTask),
          },
        ],
        task: checkLicensesTask,
      });
    }
  }
}

/**
 * Add a task to generate API docs
 *
 * @param project The project to add the task to
 */
export function setupJsiiDocgen(project: NodeProject) {
  const docgenTask = project.tasks.tryFind("docgen");

  if (docgenTask) {
    docgenTask.reset(
      "jsii-docgen .jsii -o docs/api/projen --split-by-submodule"
    );
  }

  // ignoring the entire docusaurus folder because it's not needed in the published package
  project.npmignore?.exclude("/docusaurus/");
}

/**
 * Add a task to upgrade dependencies
 *
 * @param project The project to add the task to
 */
export function setupUpgradeDependencies(project: NodeProject) {
  // Upgrade Dependencies in two parts:
  // a) Upgrade bundled dependencies as a releasable fix
  // b) Upgrade devDependencies as a chore
  new UpgradeDependencies(project, {
    taskName: "upgrade-bundled",
    types: [DependencyType.BUNDLED],
    semanticCommit: "fix",
    pullRequestTitle: "upgrade bundled dependencies",
    workflowOptions: {
      labels: ["auto-approve"],
      // Run projen's daily upgrade (and release) acyclic to the schedule that projects are on so they get updates faster
      schedule: UpgradeDependenciesSchedule.expressions(["0 12 * * *"]),
    },
  });
  new UpgradeDependencies(project, {
    taskName: "upgrade",
    exclude: [
      // exclude the bundled deps
      ...project.deps.all
        .filter((d: any) => d.type === DependencyType.BUNDLED)
        .map((d: any) => d.name),
      // constructs version constraint should not be changed
      "constructs",
      // markmac depends on projen, we are excluding it here to avoid a circular update loop
      "markmac",
    ],
    workflowOptions: {
      labels: ["auto-approve"],
    },
  });
}

/**
 * Add only the gitignore inclusions and exclusions rules that are not related to other tools
 *
 * @param project The project to add the rules to
 */
export function setupGitignore(project: Project) {
  project.gitignore.include("/test/inventory/**");
  project.gitignore.exclude("/.idea");
  project.gitignore.exclude("**/.tool-versions");
}

/**
 * Setup markdown readme macros and markdownlint
 *
 * @param project The project to add the rules to
 */
export function setupMarkdown(project: NodeProject) {
  // expand markdown macros in readme
  const macros = project.addTask("readme-macros");
  macros.exec("mv README.md README.md.bak");
  macros.exec("cat README.md.bak | markmac > README.md");
  macros.exec("rm README.md.bak");
  project.postCompileTask.spawn(macros);

  new JsonFile(project, ".markdownlint.json", {
    obj: {
      default: true,
      "commands-show-output": false,
      "line-length": {
        line_length: 200,
      },
    },
  });
  project.npmignore?.exclude("/.markdownlint.json");
}

/**
 * Setup vscode configuration
 *
 * @param project The project to add the configuration to
 */
export function setupVscode(project: NodeProject) {
  project.vscode?.launchConfiguration.addConfiguration({
    type: "pwa-node",
    request: "launch",
    name: "projen CLI",
    skipFiles: ["<node_internals>/**"],
    program: "${workspaceFolder}/lib/cli/index.js",
    outFiles: ["${workspaceFolder}/lib/**/*.js"],
  });
}

/**
 * Setup mergify rules
 * @param project The project to add the rules to
 */
export function setupMergify(project: NodeProject) {
  project.github?.mergify?.addRule({
    name: "Label core contributions",
    actions: {
      label: {
        add: ["contribution/core"],
      },
    },
    conditions: ["author~=^(eladb|Chriscbr)$", "label!=contribution/core"],
  });
}

/**
 * Setup gitpod configuration
 *
 * @param project The project to add the configuration to
 */
export function setupGitpod(project: NodeProject) {
  project.gitpod?.addCustomTask({
    name: "Setup",
    init: "yarn install",
    prebuild: "bash ./projen.bash",
    command: "npx projen build",
  });
}

/**
 * Setup devcontainer configuration
 *
 * @param project The project to add the configuration to
 */
export function setupDevcontainer(project: NodeProject) {
  const setup = project.addTask("devenv:setup");
  setup.exec("yarn install");
  setup.spawn(project.buildTask);
  project.devContainer?.addTasks(setup);

  project.npmignore?.exclude("/.devcontainer.json");
}

/**
 * Setup all-contributors
 *
 * @param project The project to add the configuration to
 */
export function setupAllContributors(project: NodeProject) {
  project.addTask("contributors:update", {
    exec: 'all-contributors check | grep "Missing contributors" -A 1 | tail -n1 | sed -e "s/,//g" | xargs -n1 | grep -v "\\[bot\\]" | grep -v "cdklabs-automation" | xargs -n1 -I{} all-contributors add {} code',
  });
  project.npmignore?.exclude("/.all-contributorsrc");
}

/**
 * Add the npmignore rules that are not related to other tools
 *
 * @param project The project to add the rules to
 */
export function setupNpmignore(project: NodeProject) {
  project.npmignore?.exclude("/docs/");
  project.npmignore?.exclude("/logo/");
  project.npmignore?.exclude("/rfcs/");
  project.npmignore?.exclude("/scripts/");
  project.npmignore?.exclude("/ARCHITECTURE.md");
  project.npmignore?.exclude("/CODE_OF_CONDUCT.md");
  project.npmignore?.exclude("/CONTRIBUTING.md");
  project.npmignore?.exclude("/VISION.md");
  project.npmignore?.exclude("/SECURITY.md");
  project.npmignore?.exclude("/.gitpod.yml");
}

export function setupBiomeTypesGeneration(project: NodeProject) {
  const genetateBiomeTypesTask = project.addTask("generate-biome-types", {
    exec: "node ./scripts/generate-biome-types.js",
  });

  const preCompileTask = project.tasks.tryFind("pre-compile");
  // Add type generation to build step to include it also to self-mutation
  if (preCompileTask) {
    preCompileTask.insertStep(preCompileTask.steps.length, {
      spawn: genetateBiomeTypesTask.name,
    });
  }
}
