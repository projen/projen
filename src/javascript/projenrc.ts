import { existsSync, mkdirSync, writeFileSync } from "fs";
import { dirname, resolve } from "path";
import { Eslint } from "./eslint";
import { renderJavaScriptOptions } from "./render-options";
import { DEFAULT_PROJEN_RC_JS_FILENAME } from "../common";
import type { Project } from "../project";
import { ProjenrcFile } from "../projenrc";
import { Biome } from "./biome/biome";

export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.js"
   */
  readonly filename?: string;
}

/**
 * A projenrc file written in JavaScript
 *
 * This component can be instantiated in any type of project
 * and has no expectations around the project's main language.
 */
export class Projenrc extends ProjenrcFile {
  public readonly filePath: string;

  constructor(project: Project, options: ProjenrcOptions = {}) {
    super(project);

    this.filePath = options.filename ?? DEFAULT_PROJEN_RC_JS_FILENAME;

    // this is the task projen executes when running `projen`
    project.defaultTask?.exec(`node ${this.filePath}`);

    this.generateProjenrc();
  }

  public override preSynthesize(): void {
    super.preSynthesize();

    const eslint = Eslint.of(this.project);
    eslint?.addLintPattern(this.filePath);
    eslint?.addIgnorePattern(`!${this.filePath}`);
    eslint?.addOverride({
      files: [this.filePath],
      rules: {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    });

    const biome = Biome.of(this.project);
    biome?.addLintPattern(this.filePath);
  }

  private generateProjenrc() {
    const rcfile = resolve(this.project.outdir, this.filePath);
    if (existsSync(rcfile)) {
      return; // already exists
    }

    const bootstrap = this.project.initProject;
    if (!bootstrap) {
      return;
    }

    const parts = bootstrap.fqn.split(".");
    const moduleName = parts[0];
    const importName = parts[1];
    const className = parts.slice(1).join(".");

    const { renderedOptions, imports } = renderJavaScriptOptions({
      comments: bootstrap.comments,
      args: bootstrap.args,
      type: bootstrap.type,
    });

    imports.add(moduleName, importName);

    const lines = new Array<string>();
    lines.push(...imports.asCjsRequire());
    lines.push();
    lines.push(`const project = new ${className}(${renderedOptions});`);
    lines.push();
    lines.push("project.synth();");

    mkdirSync(dirname(rcfile), { recursive: true });
    writeFileSync(rcfile, lines.join("\n"));
    this.project.logger.info(
      `Project definition file was created at ${rcfile}`
    );
  }
}
