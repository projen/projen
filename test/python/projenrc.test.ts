import * as path from "path";
import { renderProjenInitOptions } from "../../src/javascript/render-options";
import { ProjectType } from "../../src/project";
import { Projenrc, resolvePythonImportName } from "../../src/python/projenrc";
import { synthSnapshot, TestProject, withProjectDir } from "../util";

test("projenrc.py support", () => {
  // GIVEN
  const project = new TestProject();

  // WHEN
  new Projenrc(project, {
    projenVersion: "1.2.3",
  });

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});

test("generate projenrc in python", () => {
  // GIVEN
  const project = new TestProject(
    renderProjenInitOptions("projen.python.PythonProject", {})
  );

  // WHEN
  new Projenrc(project);

  // THEN
  expect(synthSnapshot(project)[".projenrc.py"]).toMatchSnapshot();
});

test("javascript values are translated to python", () => {
  // GIVEN
  const project = new TestProject(
    renderProjenInitOptions("projen.python.PythonProject", {
      stringArg: "hello",
      intArg: 123,
      floatArg: 123.45,
      booleanArg: false,
      undefinedArg: undefined,
      nullArg: null,
      objectArg: { foo: "bar" },
      projectType: ProjectType.LIB,
    })
  );

  // WHEN
  new Projenrc(project);

  // THEN
  expect(synthSnapshot(project)[".projenrc.py"]).toMatchSnapshot();
});

test("ensure python import is correctly resolved when python module exists", () => {
  // GIVEN
  const jsiiManifest: any = {
    targets: {
      python: {
        module: "my_module",
      },
    },
  };

  const fqnInputToExpectedImportMap = {
    "my-module.component": "my_module.component",
    "my_module.component": "my_module.component",
    "my_module.nested.component": "my_module.nested.component",
  };

  Object.entries(fqnInputToExpectedImportMap).forEach(
    ([jsiiFqn, expectedImportName]) => {
      // WHEN
      const resolvedImportName = resolvePythonImportName(jsiiFqn, jsiiManifest);

      // THEN
      expect(resolvedImportName).toEqual(expectedImportName);
    }
  );
});

test("ensure python import is correctly resolved to jsiiFqn when python module is undefined", () => {
  // GIVEN
  const jsiiManifest: any = {};
  const jsiiFqn = "my-module.component";

  // WHEN
  const resolvedImportName = resolvePythonImportName(jsiiFqn, jsiiManifest);

  // THEN
  expect(resolvedImportName).toEqual(jsiiFqn);
});

test("generate projenrc in python with a given outdir", () => {
  withProjectDir((projectdir) => {
    const newOutDir = path.join(projectdir, "foo");
    const project = new TestProject(
      renderProjenInitOptions("projen.python.PythonProject", {
        outdir: newOutDir,
      })
    );
    new Projenrc(project);
    const projen = new Projenrc(project);

    expect(projen.project.outdir).toEqual(newOutDir);
  });
});
