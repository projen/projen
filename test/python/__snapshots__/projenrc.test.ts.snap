// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate projenrc in python 1`] = `
"from projen.python import PythonProject

project = PythonProject()

project.synth()"
`;

exports[`javascript values are translated to python 1`] = `
"from projen.python import PythonProject
from projen import ProjectType

project = PythonProject(
    string_arg="hello",
    int_arg=123,
    float_arg=123.45,
    boolean_arg=False,
    undefined_arg=None,
    null_arg=None,
    object_arg={"foo":"bar"},
    project_type=ProjectType.LIB,
)

project.synth()"
`;

exports[`projenrc.py support 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".

* text=auto eol=lf
/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.projenrc.py
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".",
    "dependencies": [
      {
        "name": "projen",
        "type": "devenv",
        "version": "1.2.3",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".",
    "files": [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
        "steps": [
          {
            "exec": "python .projenrc.py",
          },
        ],
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "README.md": "# replace this",
}
`;
